effectful/handlers/llm/template.py

```diff
@@ -0,0 +1,94 @@
+import inspect
+import types
+import typing
+from collections import ChainMap
+from collections.abc import Callable, Mapping
+from typing import Any
+
+from effectful.ops.types import NotHandled, Operation
+
+
+class Tool[**P, T](Operation[P, T]):
+    def __init__(
+        self, signature: inspect.Signature, name: str, default: Callable[P, T]
+    ):
+        if not default.__doc__:
+            raise ValueError("Tools must have docstrings.")
+
+        for param_name, param in signature.parameters.items():
+            # allow self to be unannotated
+            if param_name != "self" and param.annotation is inspect.Parameter.empty:
```

Comment:
Matching on the literal `"self"` to detect that this is a descriptor is not sound, this should be fixed or removed.